### ISSUE TITLE
Split monitor into virtual outputs

### ISSUE THREAD

**Issue by aDogCalledSpot** [hyprwm/Hyprland#3575]
### Description

[Sway has already covered this issue quite extensively](https://github.com/swaywm/sway/issues/1666#issuecomment-1356871296).

The idea is that you can split an (ultrawide) monitor into multiple smaller virtual outputs.

Example: Instead of having a single 5440x1440 monitor, you could divide it into two 2560x1440 monitors or even a setup such as 1280x1440 + 2560x1440 + 1280x1440 monitor

This way you can switch between workspaces on different parts of the physical screen indepently of the rest of the screen.

If someone could give me some pointers, I would be happy to draft a PR for this.

**Comment by gernotfeichter** [comment]
I have the opposite requirement: One virtual output should span over two physical monitors. Would be awesome if the code could cover that scenario as well!

**Comment by blueStruct** [comment]
> I have the opposite requirement: One virtual output should span over two physical monitors. Would be awesome if the code could cover that scenario as well!

As I recently checked again if sway or hyprland had implemented splitting a monitor (the feature in this issue) since the last time I checked, I came across a hyprland plugin that seems to do what you want:
https://github.com/levnikmyskin/hyprland-virtual-desktops

I found it on the hyprland awesome list:
https://github.com/hyprland-community/awesome-hyprland

Oh, and, of course, a solution that is generic enough to cover both scenarios would be awesome. 

One solution I thought about was whether the fake outputs feature could already be used to split a monitor into multiple.

**Comment by aDogCalledSpot** [comment]
@vaxerski could you give a high-level overview of what would need to be done to support this, so that someone could maybe look into it a bit?

**Comment by vaxerski** [comment]
this is all but simple; hyprland is very much not made with this in mind.

**Comment by aDogCalledSpot** [comment]
Are there possible first steps that could be done in order to slowly move towards this? I dont expect this to be solved with a single PR.

**Comment by vaxerski** [comment]
none that I can think of. the entire thing is built on the assumption that there is strictly 1 monitor per 1 workspace

**Comment by Rattlyy** [comment]
Can second this, I own an ultrawide and I can't use it with hyprland.

**Comment by ZackMitkin** [comment]
An alternative is to use PBP mode, idk if there's any monitors that support 3 side by side PBP outputs though. I find it pretty clunky because the only way to center a window between 2 monitors is through floating.

I just use master layout with `always_center_master`, it works pretty well except I don't have control over resizing the non-master windows individually. I just got used to it.

**Comment by tobh** [comment]
I'm currently using XMonad use layoutScreens there to cycle through three configurations for my ultrawide (5120x1440):
- 3 virtual monitors with the middle one being the normal sized (2560x1440) and the side ones being half sized
- 2 virtual monitors with the same 2560x1440 resolution
- 1 big monitor using all available space

XMonad handles everything inside these virtual monitors as if they where real displays and it also uses the available resolution of each virtual monitor to select the best layout (e.g. the half sized displays have a master window on top and other windows below tiled and the full sized ones have the master on the left or right).

If hyprland could or does support something similar it would be great.

**Comment by endeavour** [comment]
I don't know much about wayland but a quick glance through the docs suggests to me that this is something that should sit above the compositor and maps real wl_output devices (monitors) onto multiple virtual wl_outputs? That way it wouldn't be hyprland specific? Maybe someone more knowledgeable can confirm if this is possible.

**Comment by marcSoda** [comment]
I too am unable to switch to hyprland until this is supported. I have an ultrawide that I would like to split into 2 or 3 virtual displays. It would be great if hyprland could treat it as 3 separate displays. This is how you  would do it in x11:

```
  xrandr --setmonitor LeftSide 1280/340x1440/381+1920+0 none > /dev/null 2>&1
  xrandr --setmonitor Center 2560/677x1440/381+3200+0 "$connected_monitor" > /dev/null 2>&1
  xrandr --setmonitor RightSide 1280/340x1440/381+5760+0 none > /dev/null 2>&1
```

**Comment by foodornt** [comment]
Yes please, I usually split my monitor into two: half for notes and what I'm doing (an ADHD must) and the other half is my workspace.
I wish I could make the switch but this is essential if I want to maintain productivity

**Comment by reversesh3ll** [comment]
> If someone could give me some pointers, I would be happy to draft a PR for this.

What information do you need to proceed with a PR?

**Comment by aDogCalledSpot** [comment]
I was hoping for a high level overview from someone who is more familiar with the codebase of what would need to be done.

I haven't worked on the code at all so I'm completely clueless as to what's necessary.

Right now though I wouldn't have the capacity to work on this anyway. Would still be good to have a rough outline anyway in case someone else wants to look into it (or if I find the time at some point).

**Comment by vaxerski** [comment]
making this work would require a pretty much from-the-ground-up rework of how window management works under hyprland and is not feasible (or planned) at the moment.

You can still write your own layout that would allow that, but a layout-agnostic method is not feasible atm.

**Comment by reversesh3ll** [comment]
With the recent release of the wayland workspaces protocol (I'm unsure if hyprland will even use this protocol), but could this help us progress with this feature request? I believe it should remove some of the development complexity.

I'm dying to have this feature completed. Its the only thing holding me back from switching...

**Comment by vaxerski** [comment]
the wayland protocol is in no way related to this

**Comment by vaxerski** [comment]
Hello there!

This issue has been closed, as we are moving from issues to discussions. More about this can be found [here](https://wiki.hyprland.org/Contributing-and-Debugging/Issue-Guidelines/) and in #9854.

Firstly, please make sure this issue is still relevant in latest Hyprland. If so, we ask you to open a [discussion](https://github.com/hyprwm/Hyprland/discussions) (please read the discussion guidelines first, in the pinned post)

Thank you for your understanding!

**Comment by alcxyz** [comment]
I use an ultrawide 49" and I found the best solution in a scrolling layout like hyprscrolling, but honestly even after having hacked it to work well, its still not as good as niri for the scrolling experience.. But I like the scratchpad too much to give up on (for the moment at least). I can easily split up the workspace with 3 or 4 windows (or more..), placing a big center window and having two smaller on the side (which we obviously all want with such a monitor). My nix-config is public, and can be used as reference on how I am managing this

**Comment by reversesh3ll** [comment]
For anyone interested. I created a hyprland plugin called [hyprsplitrow](https://github.com/reversesh3ll/hyprsplitrow) that can simulate splitting a screen horizontally, creating two rows of individually switched workspaces on my 48 inch monitor. If there is enough interest, I could have a go at making it also work vertically for ultrawides. It has some limitations but functionally it works perfect to replace my previous i3 setup with xrand virtual monitors workflow. Including features like pseudo/fake fullscreen for each row (virtual monitor) etc.

**Comment by reversesh3ll** [comment]
I have updated the plugin to now work with ultrawides. Can only make two virtual monitors but should cover most people. You can also configure their size or change their size with keybinds on the fly. Next update will be to allow for other layouts. (other than the current column layout)

**Comment by pwhelan** [comment]
> but should cover most people.

I think most people would want 3 splits, one wide one in the middle and two skinny ones at each side. That way the main split is centered.
